File: base/src/main/java/com/windscribe/vpn/serverlist/entity/LocationAndDatacenters.java
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.vpn.serverlist.entity;

import androidx.annotation.Keep;
import androidx.room.Embedded;
import androidx.room.Ignore;
import androidx.room.Relation;

import java.util.List;
import java.util.Objects;

@Keep
public class LocationAndDatacenters {

    @Relation(parentColumn = "region_id", entityColumn = "region_id", entity = Datacenter.class)
    private List<Datacenter> datacenters;

    @Ignore
    private int latencyTotal;

    @Embedded
    private Location location;

    public List<Datacenter> getDatacenters() {
        return datacenters;
    }

    public void setDatacenters(List<Datacenter> datacenters) {
        this.datacenters = datacenters;
    }

    @Ignore
    public int getLatencyTotal() {
        return latencyTotal;
    }

    @Ignore
    public void setLatencyTotal(int latencyTotal) {
        this.latencyTotal = latencyTotal;
    }

    public Location getLocation() {
        return location;
    }

    public void setLocation(Location location) {
        this.location = location;
    }

    @Ignore
    public Boolean isExpanded = false;

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof LocationAndDatacenters)) return false;
        LocationAndDatacenters that = (LocationAndDatacenters) o;
        return latencyTotal == that.latencyTotal && datacenters.equals(that.datacenters) && location.equals(that.location) && isExpanded.equals(that.isExpanded);
    }

    @Override
    public int hashCode() {
        return Objects.hash(datacenters, latencyTotal, location, isExpanded);
    }
}

File: base/src/main/java/com/windscribe/vpn/serverlist/entity/LocationAndDatacenters.kt
/*
 * Copyright (c) 2021 Windscribe Limited.
 */
package com.windscribe.vpn.serverlist.entity

import androidx.annotation.Keep
import androidx.room.Embedded
import androidx.room.Ignore
import androidx.room.Relation
import java.util.Objects

@Keep
class LocationAndDatacenters {
    @Relation(parentColumn = "region_id", entityColumn = "region_id", entity = Datacenter::class)
    var datacenters: List<Datacenter> = emptyList()

    @Ignore
    var latencyTotal: Int = 0

    @Embedded
    var location: Location? = null

    @Ignore
    var isExpanded: Boolean = false

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is LocationAndDatacenters) return false
        return latencyTotal == other.latencyTotal &&
            datacenters == other.datacenters &&
            location == other.location &&
            isExpanded == other.isExpanded
    }

    override fun hashCode(): Int = Objects.hash(datacenters, latencyTotal, location, isExpanded)
}
